refactor!: NetworkPrefabHandler and INetworkPrefabInstanceHandler API & XmlDoc changes

NoelStephensUnity opened the pull request:
[MTT-961](https://jira.unity3d.com/browse/MTT-961)
- Improved comments and added additional XML Documentation for the NetworkPrefabHandler and associated INetworkPrefabInstanceHandler.
- Renamed the two methods within INetworkPrefabInstanceHandler to better reflect what their function was and to avoid user confusion.
- Provided the ability for Hosts to properly display NetworkPrefab overrides.
- Now when hosts spawn the NetworkPrefab override, messages to clients about the overridden NetworkObject will send the source NetworkPrefab's GlobalObjetIdHash value as opposed to the overriden NetworkPrefab's.
- The above two refactors are applied to NetworkPrefab Handlers
- Host can now register 1 or more NetworkPrefab overrides with the NetworkPrefabHandler, which allows for the destory method of the INetworkPrefabInstanceHandler implementation to be invoked on the host.
- Removed areas where we were recursively calling OnDespawnObject where the prefab message handler was detected on a NetworkObject.
- NetworkPrefabPool has been updated to reflect some of the above changes
- Improved the GenericNetworkObjectBehaviour and adjusted it to account for some of the above changes as well as clients now will no longer have a "flicker" in re-cycled and re-spawned NetworkObjects that was due to latency in the new NetworkTransform position (I think when snapshot handles spawning and despawning this could be removed).
- Added new method NetworkManager.GetNetworkPrefabOverride that is specific to Prefab Handler Overrides, Host mode, and the above changes (although there could be other uses for this)

NoelStephensUnity committed:
refactor and style

Renamed the two methods within INetworkPrefabInstanceHandler to better reflect what their function was and to avoid user confusion.

Improved comments and added additional XML Documentation for the NetworkPrefabHandler and associated INetworkPrefabInstanceHandler

NoelStephensUnity committed:
style

Minor updates on the ClientInstantiateOverride parameter names

0xFA11 commented on com.unity.multiplayer.mlapi/Runtime/Spawning/NetworkPrefabHandler.cs:
```diff
@@ -4,13 +4,36 @@
+        /// <param name="position">the initial/default position for the <see cref="NetworkObject"/> to be instantiated</param>
+        /// <param name="rotation">the initial/default rotation for the <see cref="NetworkObject"/> to be instantiated</param>
+        /// <returns></returns>
+        NetworkObject ClientInstantiateOverride(ulong ownerClientId, Vector3 position, Quaternion rotation);
+
+        /// <summary>
+        /// Client and Server
+        /// Once an implementation is registered with the <see cref="NetworkPrefabHandler"/>, this method will be called every time
+        /// a Network Prefab associated <see cref="NetworkObject"/> is destroyed.
+        /// If <see cref="NetworkObject.Despawn(bool)"/> is invoked with the default destroy parameter (i.e. false) then this method will not be invoked.
+        ///
+        /// Note on Pooling: When invoked, you do not need to despawn or destroy the NetworkObject as long as you want your pool to persist.
+        /// The most common approach is to make the <see cref="NetworkObject"/> inactive by calling <see cref="GameObject.SetActive(bool)"/>.
+        /// </summary>
+        /// <param name="networkObject">The <see cref="NetworkObject"/> being destroyed</param>
+        void DestroyOverride(NetworkObject networkObject);
```
I'm not sure if I'd go for `ClientInstantiateOverride` & `DestroyOverride` instead of `HandleNetworkPrefabSpawn` & `HandleNetworkPrefabDestroy` under an interface called `INetworkPrefabInstanceHandler`.

I think I'd drop "-Override" suffix and keep "Handle-" prefix around.

Also "Client-" prefix sounds a little risky to me in terms of our future potential and its API-breaking nature.
We might expand this feature's impact area and even make it work for both client and server sides. That would require another API-breaking change just to drop "Client-" prefix or we might have to add another super similar API that starts with "Server-" prefix to not break the old API (which would not be ideal IMHO).

Only thing I'd change here would be `HandleNetworkPrefabDestroy` → `HandleNetworkPrefabDespawn` to match `NetworkBehaviour.OnNetworkSpawn()` and `NetworkBehaviour.OnNetworkDespawn()` naming convention.

Also, it'd be tempting to keep client-specific comments around but I think they should go away and make feature non-client specific. Client-specific comment may go under `NetworkPrefabHandler` since it is the source where client-specific nature is coming from (and even that might change in the future to cover both client and server side, hence my earlier comments above).

LukeStampfli replied:
So the old naming here was very inconsistent. `Spawn`, `Destroy` don't match. We either want `HandleNetworkPrefabDespawn` or rename the `Spawn` to `Instantiate`.

Now the reason we want for `Instantiate/Destroy` was is because you can spawn/despawn a networkobject in MLAPI without triggering these handlers (GameObject.Destroy on the server for instance skips the handler). And what they really do is override a GameObject `Instatiate` / `Destroy` function and replace it with something else.

I prefer the `Override` suffix over `Handle` prefix. Because `HandleNetworkPrefabDestroy` does not indicate whether this handler is run in addition to the destroy or instead of the destroy. Dropping the `Client` prefix on `InstantiateOverride` sounds good if we consider future potential.

0xFA11 replied:
so it's:

- `HandleNetworkPrefabSpawn`
- `HandleNetworkPrefabDespawn`

vs

- `InstantiateOverride`
- `DestroyOverride `

:)

I believe it's fair to say that we'd go for `Instantiate`/`Destroy`.

so my final suggestion would be:

- `HandleInstantiate`
- `HandleDestroy`

:P

LukeStampfli replied:
HandleInstantiate, HandleDestroy sounds good to me 👍

NoelStephensUnity replied:
The only issue I have with "HandleInstantiate" is that this is only invoked on the client side and could be misleading.  However, that can be made clear in documentation.

How about we even further simplify this?  I mean it is a "Handler" so not sure we have to put "handle" in front of all of the methods.  Why not:

- Instantiate

- Destroy

0xFA11 replied:
> this is only invoked on the client side

hence my first comment, it is currently client-only but we could expand it to cover both server and client.

> it is a "Handler" so ...

I think that's also a fair point. `INetworkPrefabInstanceHandler.Instantiate` & `INetworkPrefabInstanceHandler.Destroy` it is :)

NoelStephensUnity committed:
Merge branch 'develop' into refactor/INetworkPrefabInstanceHandler

NoelStephensUnity committed:
refactor and style

Renamed the interface methods.  Updated comments.
Refactored NetworkPrefabHandler to handle GitHub Issue 990 as well as made adjustments to be able to send the appropriate GlobalObjectIdHash as a host.
This also includes a public facing method in the NetworkManager to be able to lookup a NetworkManager defined NetworkPrefab Override for some associated cases relating to Host mode and NetworkPrefabHandlers and NetworkPrefab Overrides.
Made adjustments to the SceneTransitioning test and associated scenes and assets.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/INetworkPrefabInstanceHandler

0xFA11 commented on testproject/Assets/Tests/Manual/Scripts/GenericNetworkObjectBehaviour.cs:
```diff
@@ -86,8 +86,7 @@ private void OnTriggerEnter(Collider other)
                 else
                 {
-                    NetworkObject.Despawn();
-                    NetworkObject.gameObject.SetActive(false);
```
did we drop this intentionally?
I'm talking about `NetworkObject.gameObject.SetActive(false);`

NoelStephensUnity replied:
Review it again as I just updated.
It should make more sense...

0xFA11 commented on testproject/Assets/Tests/Manual/Scripts/NetworkPrefabPool.cs:
```diff
@@ -85,6 +82,21 @@ private void RegisterCustomPrefabHandler()
         }
 
+
+        private void DeRegisterCustomPrefabHandler()
```
super-nit :)
```suggestion
        private void DeregisterCustomPrefabHandler()
```

NoelStephensUnity committed:
refactor and style

Added the case where no prefab handler override is registered but there is still a host.
Added a lookup table to quickly get the proper source GlobalObjectIdHash value for the NetworkPrefab that has an override in the NetworkManager.
The first scene in the SceneTransitioningTest uses a NetworkManager defined prefab override, in host mode the host will pool while clients will not pool (no need since there is no way to override the instantiation) and the host will always send the source NetworkPrefab's GlobalObjectIdHash value as opposed to the target override's GlobalObjectIdHash.
Fixed Fatih's super-nit-pick :)

NoelStephensUnity committed:
fix

Removing an artifact property from a previous merge.

NoelStephensUnity committed:
fix

Fixing missing change to deregister and issue with despawning during fixed update.

NoelStephensUnity committed:
fix

Need to reset the position for the object when no handler exists otherwise the trigger will fire immediately and cause it to despawn immediately.

NoelStephensUnity committed:
refactor and fix

Removing despawn replication and from places where they are no longer needed either due to the now proper message ordering.

Updated the GenericNetowrkObjectBehaviour such that clients disable the MeshRenderer component and then have a delayed re-enabling of it upon being spawned in order to avoid the latent update in the transform's position (removes the flickering of NetworkObjects on the client).

Added the ability to make the SwitchSceneHandler have a delayed "auto switch" in order to run cycled long burn runs to assure everything is working "as expected".

Fixed an access violation in StatsDisplay ( minor thing that was bugging me).

0xFA11 reviewed (changes requested):
I'm seeing a few other changes going in after I approved this PR.
I think I should re-review this again some time — please hmu when you think it's ready to be reviewed again :)

NoelStephensUnity commented on com.unity.multiplayer.mlapi/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -504,14 +504,14 @@ internal void ServerDestroySpawnedSceneObjects()
                     // that occur shortly after NetworkSceneManager invokes ServerDestroySpawnedSceneObjects
                     // within the NetworkSceneManager.SwitchScene method.
-                    SpawnedObjectsList.Remove(sobj);
+
```
This is really a short term change and will be updated in PR-955 Additive Scene Loading.

NoelStephensUnity commented on com.unity.multiplayer.mlapi/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -559,7 +559,10 @@ internal void DestroySceneObjects()
                         {
                             NetworkManager.PrefabHandler.HandleNetworkPrefabDestroy(networkObjects[i]);
-                            OnDespawnObject(networkObjects[i], false);
+                            if (SpawnedObjects.ContainsKey(networkObjects[i].NetworkObjectId))
```
More short term updates until PR-955.

NoelStephensUnity committed:
fix and refactor

Refactored a better client side "no blinking when spawning" solution.

Fixed an issue where everyone was sending destroy messages to everyone else.

NoelStephensUnity commented on com.unity.multiplayer.mlapi/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -504,14 +504,14 @@ internal void ServerDestroySpawnedSceneObjects()
+
                     if (NetworkManager.PrefabHandler != null && NetworkManager.PrefabHandler.ContainsHandler(sobj))
                     {
                         NetworkManager.PrefabHandler.HandleNetworkPrefabDestroy(sobj);
```
This I am leaving up to the user to handle how they want to deal with scene transitions and prefab handler overrides.

NoelStephensUnity commented on com.unity.multiplayer.mlapi/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -681,7 +684,6 @@ internal void OnDespawnObject(NetworkObject networkObject, bool destroyGameObjec
                 if (NetworkManager.PrefabHandler.ContainsHandler(networkObject))
                 {
                     NetworkManager.PrefabHandler.HandleNetworkPrefabDestroy(networkObject);
-                    OnDespawnObject(networkObject, false);
```
@LukeStampfli , I could you double check this?  I did some testing on my end and I think this is no longer needed with the more recent updates (I think this was for some message ordering issue, but that should be fixed now).
Gladly will revert this is you think there is a special case for having that OnDespawnObject here.

NoelStephensUnity commented on com.unity.multiplayer.mlapi/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -681,7 +684,6 @@ internal void OnDespawnObject(NetworkObject networkObject, bool destroyGameObjec
                 if (NetworkManager.PrefabHandler.ContainsHandler(networkObject))
                 {
                     NetworkManager.PrefabHandler.HandleNetworkPrefabDestroy(networkObject);
-                    OnDespawnObject(networkObject, false);
```
@LukeStampfli , I could you double check this?  I did some testing on my end and I think this is no longer needed with the more recent updates (I think this was for some message ordering issue, but that should be fixed now).
Gladly will revert this is you think there is a special case for having that OnDespawnObject here.

LukeStampfli replied:
I tried understand this but not sure why we had it. It does a recursive call into OnDespawnObject again for some reason I can't find a reason why we would want to do that. So from what I know this is not needed?

NoelStephensUnity committed:
refactor

Added a more simplified example that uses a INetworkPrefabInstanceHandler implementation as a 1:1 NetworkObject pool.

LukeStampfli reviewed (approved):
Looks all good and great XML docs!

NoelStephensUnity committed:
fix

Fixed issue with spawning:
- Sending to every client for each client within the observers list
- Making the client ownership equal to the client it was sending to

This fixes the issue with connecting more than 1 client to a host or server.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/INetworkPrefabInstanceHandler

LukeStampfli reviewed (approved):
The additional fix to the `SendSpawnCallForObject` looks good.
